# Conflict And Duplication Identifier (CAnDI)

scripts to investigate gene tree conflict with gene families

**Python 3 version:** The program has also been converted to python3 and can be found in the folder Py3version. Although this version has been tested, there still may be some issues that arose from converting the program, and the python2 version used for the analyses in the paper is still the main one in the scripts folder for the time being.

## Introduction

This repository contains the CAnDI.py program designed to detect conflict using orthologs (orthologous gene trees) or gene families (Homologous gene trees). The program can either map a series of gene trees onto the species tree to identify the amount of conflict the gene trees have, map a species onto a gene tree to investigate where the gene tree has conflict, or search for patterns of conflict. A more detailed description of all the programs features may be found in the manual.

License: GPL https://www.gnu.org/licenses/gpl-3.0.html

## Publications
The MS describing CAnDI is not yet published, but is available on BioRXiv at https://doi.org/10.1101/2023.11.18.567661. 

In addition, CAnDI has been used to conduct conflict analysis in one peer-reviewed paper: https://doi.org/10.1093/molbev/msac044. 
